marker_rustc_driver/src/conversion/marker/common/span.rs

```diff
@@ -0,0 +1,70 @@
+use marker_api::{
+    ast::{ExpnInfo, FileInfo, FilePos, SpanPos, SpanSource},
+    prelude::Span,
+};
+
+use crate::conversion::marker::MarkerConverterInner;
+
+impl<'ast, 'tcx> MarkerConverterInner<'ast, 'tcx> {
+    pub fn to_span(&self, rustc_span: rustc_span::Span) -> Span<'ast> {
+        Span::new(
+            self.to_span_src_id(rustc_span.ctxt()),
+            // The driver resugars all expressions and spans, this should therefore
+            // only be true for spans from macro expansion.
+            rustc_span.from_expansion(),
+            self.to_span_pos(rustc_span.lo()),
+            self.to_span_pos(rustc_span.hi()),
+        )
+    }
+
+    pub fn to_span_pos(&self, byte_pos: rustc_span::BytePos) -> SpanPos {
+        SpanPos::new(byte_pos.0)
+    }
+
+    pub fn to_span_source(&self, rust_span: rustc_span::Span) -> SpanSource<'ast> {
+        let ctxt = rust_span.ctxt();
+        if ctxt.is_root() {
+            let src_file = self.rustc_cx.sess.source_map().lookup_source_file(rust_span.lo());
+            let name = match &src_file.name {
+                rustc_span::FileName::Real(
+                    rustc_span::RealFileName::LocalPath(file_path)
+                    | rustc_span::RealFileName::Remapped {
+                        virtual_name: file_path,
+                        ..
+                    },
+                ) => file_path.to_string_lossy().to_string(),
```

Comment:
This was acutally the lint I wrote as part of https://github.com/rust-marker/marker/issues/189 (`to_string_on_cow_str`).

You should never write `to_string_lossy().to_string()` write this instead, because it consumes the `Cow` and doesn't allocate when it already owned:

```suggestion
                ) => file_path.to_string_lossy().into_owned(),
```

Reply:
I still think this should be a Clippy lint, as it's generally applicable

Reply:
Regarding the `to_owned/to_string`. I don't care too much about this, but it's better to use a consistent style anyway. I know there is a clippy lint for forcing the usage of [`to_owned`](https://rust-lang.github.io/rust-clippy/rust-1.56.0/index.html#str_to_string) over `to_string`. If you don't agree with that clippy lint, we can have a custom marker lint to force `to_string` in this repo. Anyway it's better to have some automation to force a consistent style for that and prevent the argument